11-payment-encoding.md

```diff
@@ -216,7 +218,8 @@ A reader:
     - MUST use that as [`payment_secret`](04-onion-routing.md#tlv_payload-payload-format)
   - if the `c` field (`min_final_cltv_expiry`) is not provided:
     - MUST use an expiry delta of at least 18 when making the payment
-
+  - if an `m` field is provided:
+    - MUST use that as [`payment_metadata`](04-onion-routing.md#tlv_payload-payload-format)
```

Comment:
What happens if the payer does not use this as [`payment_metadata`](04-onion-routing.md#tlv_payload-payload-format) in the `tlv_payload`? 

In [BOLT 04 we have the onion failure message](https://github.com/joostjager/lightning-rfc/blob/custom-invoice-data/04-onion-routing.md#failure-messages)  1. type: PERM|15 (`incorrect_or_unknown_payment_details`)


https://github.com/joostjager/lightning-rfc/blob/custom-invoice-data/04-onion-routing.md#L953 currently reads: 

> The `payment_hash` is unknown to the final node, the `payment_secret` doesn't
> match the `payment_hash`, the amount for that `payment_hash` is incorrect or
> the CLTV expiry of the htlc is too close to the current block height for safe
> handling.

Maybe we should also include that this error is sent if the `payment_metadata` was not included?

Reply:
Or we can use `invalid_onion_payload`?

Reply:
I am a bit confused right now. Which of the two error codes protects the privacy of the recipient better?

I guess if an intermediate malicious node interrupts routing and sends onions to various candidate recipients without including `payment_metadata` those wrong candidates would regularly send `incorrect_or_unknown_payment_details` back to the intermediate node. However following your suggestion the real recipient would send back `invalid_onion_payload` as they expected the `payment_metadata` field. Thus my feeling is that `incorrect_or_unknown_payment_details` will be more secure as it prevents such an attack. Following #627 I am under the impression that `invalid_onion_payload` is intended to be used if something goes wrong for a routing node when decoding the onion. But maybe I missed something and @t-bast can clarify?

Reply:
> Thus my feeling is that incorrect_or_unknown_payment_details will be more secure as it prevents such an attack.

I agree. If you make `payment_metadata` mandatory and if it wasn't included in the onion you receive, I would send back `incorrect_or_unknown_payment_details`. You mandated the payment to include a `payment_metadata` and it didn't, so it's really an incorrect payment (and gives the best anonymity set as it's the error used whenever something is invalid with the payment at the recipient's end).

Reply:
Updated PR to return `incorrect_or_unknown_payment_details`.

Reply:
Dear @joostjager I am very pleased that you found my review & suggestion useful. Yet I am a bit surprised that you did not accept the patch that I provided 5 days ago with my suggestion and for your convenience directly as a pull request to your branch at https://github.com/joostjager/lightning-rfc/pull/1. I hope the reason is that you have simply overseen my pull request and patch. As you did not credit my contribution in your recent commit may I ask you kindly to revert the most recent addition and accept the provided patch instead? Thank you very much!

Reply:
@lightning-developer, No I didn't see your patch on my personal repo.

Generally speaking and also in this particular case, lots of people contribute ideas. I think it will get complicated to list all of them as co-authors. They may not all have contributed text literally, but an idea often deserves credit too.